components/service/glean/src/main/java/mozilla/components/service/glean/LabeledMetricType.kt

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean
+
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This implements the developer facing API for labeled metrics.
+ *
+ * Instances of this class type are automatically generated by the parsers at build time,
+ * allowing developers to record values that were previously registered in the metrics.yaml file.
+ *
+ * Unlike most metric types, LabeledMetricType does not have its own corresponding storage engine,
+ * but records metrics for the underlying metric type T in the storage engine for that type.  The
+ * only difference is that labeled metrics are stored with the special key `$category.$name/$label`.
+ * The |StorageEngineManager.collect| method knows how to pull these special values back out of the
+ * individual storage engines and rearrange them correctly in the ping.
+ */
+
+data class LabeledMetricType<T>(
+    override val disabled: Boolean,
+    override val category: String,
+    override val lifetime: Lifetime,
+    override val name: String,
+    override val sendInPings: List<String>,
+    val subMetric: T,
+    val labels: Set<String>? = null
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = (
+        subMetric as CommonMetricData).defaultStorageDestinations
+
+    private val logger = Logger("glean/LabeledMetricType")
+
+    companion object {
+        private const val MAX_LABELS = 16
+        private const val OTHER_LABEL = "__other__"
+        private val labelRegex = Regex("^[a-z_][a-z0-9_]{0,29}$")
+    }
+
+    private val seenLabels: MutableSet<String> = mutableSetOf()
+
+    /**
+     * Handles the label in the case where labels are predefined.
+     *
+     * If the given label is not in the predefined set of labels, returns [OTHER_LABEL], otherwise
+     * returns the label verbatim.
+     *
+     * @param label The label, as specified by the user
+     * @param adjusted label, possibly set to [OTHER_LABEL]
```

Comment:
Hey Mike, sorry to nitpick this, but I think that this line should be ` @returns adjusted label, possibly set to [OTHER_LABEL]` (`@returns` rather than `@param`)